bestie/server/bigquery_metrics.go

```diff
@@ -0,0 +1,288 @@
+	}
+	fmt.Fprintf(w, "Created dataset %s\n", t.formatDatasetId())
+	return nil
+}
+
+// Create a BigQuery table with a predefined schema.
+func createTable(w io.Writer, t bigQueryTable, session bigQuerySession) error {
+	// Always attempt to create the dataset that holds the table.
+	// This is faster than first querying the table metadata,
+	// then creating the dataset if not present.
+	// We expect an "already exists" error if the dataset currently
+	// exists for the project, which is the normal case.
+	if err := createDataset(w, t, databaseLocation, session); err != nil {
+		if !strings.Contains(strings.ToLower(err.Error()), "already exist") {
+			return err
+		}
+	}
+
+	// Create the table within the dataset.
+	tableSchema := getMetricsTableSchema()
+	fmt.Fprintln(w, "Metrics Table Schema:")
+	for _, field := range tableSchema {
+		fmt.Fprintf(w, "  %s (%s)\n", field.Name, field.Type)
+	}
+	fmt.Fprintln(w)
+
+	// NOTE: Not setting an expiration for BigQuery tables created by the endpoint.
+	// If an expiration was used and someone forgets to extend the deadline,
+	// the table and all of its data are deleted by BigQuery without warning.
+	metaData := &bq.TableMetadata{
+		Schema: tableSchema,
+		//ExpirationTime: time.Now().AddDate(1, 0, 0), // Table will be automatically deleted in 1 year
+	}
+	tableRef := session.client.Dataset(t.dataset).Table(t.tablename)
+	if err := tableRef.Create(session.ctx, metaData); err != nil {
+		if !strings.Contains(strings.ToLower(err.Error()), "already exist") {
+			return fmt.Errorf("Error creating table '%s': %s\n", t.formatTableId(), err)
+		}
+	} else {
+		fmt.Fprintf(w, "Created table %s\n", t.formatTableId())
+	}
+	return nil
+}
+
+// Translate protobuf metric to BigQuery metric.
+func translateMetric(stream bazelStream, m testMetric) (bigQueryMetric, error) {
+	// Process the slice of key/value pairs representing individual metric tags.
+	dat := make(map[string]string)
+	for _, kv := range m.tags {
+		dat[kv.key] = kv.value
+	}
+	// Insert additional tags using information that is only available
+	// to the BES endpoint through the Bazel event message itself.
+	//
+	// Storing both the invocationId and the invocationSha in the database,
+	// the former being useful to match up with build log references, etc.
+	dat["invocation_id"] = stream.invocationId
+	dat["invocation_sha"] = stream.invocationSha
+	dat["run"] = stream.run
+	dat["test_name"] = stream.testName
+	tags, err := json.Marshal(dat)
+	if err != nil {
+		return bigQueryMetric{}, fmt.Errorf("Error converting JSON to string: %s", err)
+	}
+
+	// Change the metric timestamp from integer to formatted string,
+	// since that is what BigQuery expects for a TIMESTAMP column.
+	//
+	// The incoming timestamp is UTC nanoseconds since epoch, but
+	// storing with microsecond granularity in BigQuery.
+	ts := m.timestamp / 1000
+	tsf := time.Unix(ts/1000000, ts%1000000).Format("2006-01-02 15:04:05.000000")
```

Comment:
What if `i.timestamp` is a raw `*time.Time`? Does the client lib do the right thing? (avoids this translation code)

Reply:
I looked into this for quite some time and concluded that we still need to do the translation. I cleaned up the code a little, copying a small time.UnixMicro() function from a more recent version of Go, but ultimately the .Format() is required to conform the TIMESTAMP string to what BigQuery requires. For example, time.UnixMicro().String() creates a nicely formatted timestamp, but it ends with "+0000 UTC", which is enough to prevent BigQuery from adding the metrics to the table. Hence, producing a custom format acceptable to BigQuery requires a .Format().